library/src/main/java/com/liulishuo/filedownloader/download/DownloadStatusCallback.java

```diff
@@ -447,20 +446,23 @@ private void handleError(Exception exception) {
         onStatusChanged(FileDownloadStatus.error);
     }
 
-    private boolean isFirstCallback = true;
+    private final AtomicBoolean isFirstCallback = new AtomicBoolean(true);
 
-    private boolean isNeedCallbackToUser(final long now) {
-        if (isFirstCallback) {
-            isFirstCallback = false;
-            return true;
+    private void inspectNeedCallbackToUser(final long now) {
+        final boolean needCallback;
+        if (isFirstCallback.compareAndSet(true, false)) {
+            needCallback = true;
+        } else {
+            final long callbackTimeDelta = now - lastCallbackTimestamp;
+            needCallback = (callbackMinIntervalBytes != NO_ANY_PROGRESS_CALLBACK
+                    && callbackIncreaseBuffer.get() >= callbackMinIntervalBytes)
+                    && (callbackTimeDelta >= callbackProgressMinInterval);
+        }
+        if (needCallback && needCallbackProgressToUser.compareAndSet(false, true)) {
+            FileDownloadLog.i(this, "inspectNeedCallbackToUser need callback to user");
+            lastCallbackTimestamp = now;
```

Comment:
These fields should be updated here to ensure synchronization, otherwise, the `handleProgress` method will be invoked more than once in a very short period of time.